REGRESSION (251454@main): Setting scrollTop on fixed element with overflow breaks scrolling on a 'overscroll-behavior:none' page

nmoucht opened the pull request:
#### b08436732d9d451b70e9d6cd5061d65d6fd54497
<pre>
REGRESSION (251454@main): Setting scrollTop on fixed element with overflow breaks scrolling on a &apos;overscroll-behavior:none&apos; page
<a href="https://bugs.webkit.org/show_bug.cgi?id=245300">https://bugs.webkit.org/show_bug.cgi?id=245300</a>
&lt;rdar://100057532&gt;

Reviewed by Simon Fraser.

This fixes a few issues with <a href="https://commits.webkit.org/251454@main.">https://commits.webkit.org/251454@main.</a> The first is the fix was originally only intended for
scroll-behavior: smooth, so only add the scrollable area when we are doing an animated scroll. The second is to be more
strict when these non-user scrollable areas are in the frame view&apos;s list of sccrollable areas. We accomplish this by
adding a call when the animation completes to updateScrollableAreaSet, which will determine if the particular scrollable
area should be in the set after the animation completes. Finally, we add a check to absoluteEventTrackingRegionsForFrame(),
to see if the scrollable area was added only because it needed to do an animation.

* LayoutTests/fast/scrolling/mac/smooth-scroll-fixed-element-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/smooth-scroll-fixed-element.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::runScrollSteps):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::animatedScrollDidEndForNode):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::ScrollAnimator::didStopAnimatedScroll):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::animatedScrollDidEnd):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollToOffset):
(WebCore::RenderLayerScrollableArea::registerScrollableAreaForAnimatedScroll):
(WebCore::RenderLayerScrollableArea::animatedScrollDidEnd):
(WebCore::RenderLayerScrollableArea::registerScrollableArea): Deleted.
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: <a href="https://commits.webkit.org/257665@main">https://commits.webkit.org/257665@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/891ab0db21a5ef7440097d69578d608b5d5ab60d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8771 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32691 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108958 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169195 "Failed to checkout and rebase branch from PR 4810") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103595 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9377 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86075 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92071 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106875 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105373 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7074 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/90591 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [💥 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34035 "An unexpected error occured. Recent messages:Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Updated gtk dependencies; Pull request contains relevant changes; Skipped layout-tests; layout-tests (exception)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21948 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2609 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23462 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2544 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45852 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8699 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42935 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4404 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2698 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

smfr reviewed (changes requested):
With this change, we'll still get a scrolling "dead zone" while the scroll animation is running, right? You could test this by making the animation run much longer.

A better approach would be to fix the non-fast scrollable region generation to ignore ScrollableAreas that were only registered for the purpose of driving their animations.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
```diff
@@ -126,7 +126,7 @@ EventTrackingRegions ScrollingCoordinator::absoluteEventTrackingRegionsForFrame(
         for (auto& scrollableArea : *scrollableAreas) {
             // Composited scrollable areas can be scrolled off the main thread.
-            if (!scrollableArea->isVisibleToHitTesting() || scrollableArea->usesAsyncScrolling())
+            if (!scrollableArea->isVisibleToHitTesting() || scrollableArea->usesAsyncScrolling() || scrollableArea->scrollAnimationStatus() == ScrollAnimationStatus::Animating)
```
Does this do the right thing for a scrollable area that is animating, but also has wheel event handlers so needs to be on the slow path?

I'd like to see more tests here.

nmoucht replied:
I don't think we have a great mechanism forcing a check to happen during the animation. I changed this check to now look at whether the scrollable area was added only for servicing the animation, so I don't think we need to worry about this anymore.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
```diff
@@ -126,7 +126,7 @@ EventTrackingRegions ScrollingCoordinator::absoluteEventTrackingRegionsForFrame(
         for (auto& scrollableArea : *scrollableAreas) {
             // Composited scrollable areas can be scrolled off the main thread.
-            if (!scrollableArea->isVisibleToHitTesting() || scrollableArea->usesAsyncScrolling())
+            if (!scrollableArea->isVisibleToHitTesting() || scrollableArea->usesAsyncScrolling() || scrollableArea->needsAnimatedScroll())
```
Isn't this still going to cause animated-scrolling ScrollableAreas to be missing from the non-fast event region while the animation is running, which means they won't get WheelEvents?

nmoucht replied:
Removed this extra condition since we now use another set to hold the non user scrollable areas.

smfr commented on Source/WebCore/platform/ScrollAnimator.cpp:
```diff
@@ -287,6 +287,7 @@ void ScrollAnimator::willStartAnimatedScroll()
 void ScrollAnimator::didStopAnimatedScroll()
 {
     m_scrollableArea.setScrollAnimationStatus(ScrollAnimationStatus::NotAnimating);
+    m_scrollableArea.animatedScrollDidEnd();
```
Seems like `setScrollAnimationStatus()` could take care of calling `animatedScrollDidEnd()`

nmoucht replied:
I think we only want to call this here since this is the call for specifically an animated scroll finishing.

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -394,6 +395,7 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea> {
     bool shouldBlockScrollPropagation(const FloatSize&) const;
     FloatSize deltaForPropagation(const FloatSize&) const;
     WEBCORE_EXPORT virtual float adjustVerticalPageScrollStepForFixedContent(float step);
+    virtual bool needsAnimatedScroll() const { return false; }
```
This isn't "needs", it's more like "isRunningAnimatedScroll" I think. Why isn't this deducible from `ScrollAnimationStatus` already?

nmoucht replied:
This bool is only set for this case so its not deducible from ScrollAnimationStatus.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -1941,4 +1941,12 @@ void RenderLayerScrollableArea::didStartScrollAnimation()
 
+void RenderLayerScrollableArea::animatedScrollDidEnd()
+{
+    if (m_needsAnimatedScroll) {
```
Don't we need similar changes for animated scrolls in `<iframe>`s?

nmoucht replied:
Per our discussion and the test I don't think so.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
```diff
@@ -315,8 +313,10 @@ ScrollOffset RenderLayerScrollableArea::scrollToOffset(const ScrollOffset& scrol
+        registerScrollableAreaForAnimatedScroll();
         ScrollableArea::scrollToPositionWithAnimation(snappedPosition);
+    }
     else if (!requestScrollPositionUpdate(snappedPosition, options.type, options.clamping))
```
Closing brace goes on this line

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -286,6 +289,8 @@ class RenderLayerScrollableArea final : public ScrollableArea {
     bool m_containsDirtyOverlayScrollbars { false };
     bool m_updatingMarqueePosition { false };
+    
+    bool m_needsAnimatedScroll { false };
```
`m_isRegisteredForAnimatedScroll`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -4454,8 +4454,13 @@ void Document::runScrollSteps()
+            auto scrollableAreasToUpdate = *frameView->scrollableAreas();
+            if (auto nonUserScrollableAreas = frameView->scrollableAreasForAnimatedScroll()) {
+                for (auto* area : *nonUserScrollableAreas)
+                    scrollableAreasToUpdate.add(area);
```
Can you do `scrollableAreasToUpdate.add(nonUserScrollableAreas.begin(), nonUserScrollableAreas.end())`?

nmoucht replied:
Fixed.

nmoucht committed:
REGRESSION (251454@main): Setting scrollTop on fixed element with overflow breaks scrolling on a 'overscroll-behavior:none' page
https://bugs.webkit.org/show_bug.cgi?id=245300
<rdar://100057532>

Reviewed by Simon Fraser.

This fixes a few issues with https://commits.webkit.org/251454@main. The first is the fix was originally only intended for
scroll-behavior: smooth, so only add the scrollable area when we are doing an animated scroll. The second is to be more
strict when these non-user scrollable areas are in the frame view's list of sccrollable areas. We accomplish this by
adding a call when the animation completes to updateScrollableAreaSet, which will determine if the particular scrollable
area should be in the set after the animation completes. Finally, we add a check to absoluteEventTrackingRegionsForFrame(),
to see if the scrollable area was added only because it needed to do an animation.

* LayoutTests/fast/scrolling/mac/smooth-scroll-fixed-element-expected.txt: Added.
* LayoutTests/fast/scrolling/mac/smooth-scroll-fixed-element.html: Added.
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::runScrollSteps):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::animatedScrollDidEndForNode):
* Source/WebCore/page/scrolling/ScrollingCoordinator.cpp:
(WebCore::ScrollingCoordinator::absoluteEventTrackingRegionsForFrame const):
* Source/WebCore/platform/ScrollAnimator.cpp:
(WebCore::ScrollAnimator::didStopAnimatedScroll):
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::animatedScrollDidEnd):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::scrollToOffset):
(WebCore::RenderLayerScrollableArea::registerScrollableAreaForAnimatedScroll):
(WebCore::RenderLayerScrollableArea::animatedScrollDidEnd):
(WebCore::RenderLayerScrollableArea::registerScrollableArea): Deleted.
* Source/WebCore/rendering/RenderLayerScrollableArea.h:

Canonical link: https://commits.webkit.org/257665@main
